Update incoming webhook for github apps

savitaashture opened the pull request:
If GithubApp is installed on the cluster then no need to provide `git_provider` spec because for GithubApps we detect the secret and use it

Signed-off-by: Savita Ashture <sashture@redhat.com>

<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

- [ ] ♽  Run `make test lint` before submitting a PR (ie: with [pre-commit](https://pipelinesascode.com/dev/tools), no need to waste CPU cycle on CI
- [ ] 📖 If you are adding a user facing feature or make a change of the behavior, please verify that you have documented it
- [ ] 🧪 100% coverage is not a target but most of the time we would rather have a unit test if you make a code change.
- [ ] 🎁 If that's something that is possible to do please ensure to check if we can add a e2e test.
- [ ] 🔎 If there is a flakiness in the CI tests then don't *necessary* ignore it, better get the flakyness fixed before merging or if that's not possible there is a good reason to bypass it. (token rate limitation may be a good reason to skip).

savitaashture committed:
Enhance incoming webhook for github apps
If GithubApp is installed on the cluster then no need to provide git_provider spec because for GithubApps we detect the secret and use it

Signed-off-by: Savita Ashture <sashture@redhat.com>

chmouel commented:
/retest

chmouel commented on pkg/adapter/incoming.go:
```diff
@@ -95,20 +221,73 @@ func (l *listener) processIncoming(targetRepo *v1alpha1.Repository) (provider.In
+}
+
+func GenerateJWT(ctx context.Context, run *params.Run) (string, error) {
+	ns := os.Getenv("SYSTEM_NAMESPACE")
```
we already have GetappToken doing this, can we not have duplication? pkg/provider/github/parse_payload.go

savitaashture replied:
Done
modified to avoid duplication

chmouel commented on pkg/adapter/incoming.go:
```diff
@@ -3,12 +3,23 @@ package adapter
 	"crypto/subtle"
+	"encoding/json"
 	"fmt"
+	"github.com/golang-jwt/jwt"
```
please use the versioned version adn the same version as what we have on ghinstallaton

```jwt "github.com/golang-jwt/jwt/v4"```

savitaashture replied:
Updated

chmouel commented on pkg/adapter/incoming.go:
```diff
@@ -86,6 +121,97 @@
+		apiURL = enterpriseURL
+	}
+
+	res, err := getResponse(ctx, http.MethodGet, installationURL, JWTtoken, run)
```
go-github should have all of this ie https://github.com/chmouel/pipelines-as-code/blob/rework-documentation/vendor/github.com/google/go-github/v48/github/apps.go#L181 can we remove this code and use the library code instead ?

savitaashture replied:
I tried to use library code  but client is getting nil https://github.com/chmouel/pipelines-as-code/blob/rework-documentation/vendor/github.com/google/go-github/v48/github/apps.go#L187 because of no installtion ID which we use  in parse_payload something like this https://github.com/openshift-pipelines/pipelines-as-code/blob/main/pkg/provider/github/parse_payload.go#L47


I mean i understood that Installtion ID is needed 
Correct me if i misunderstand

chmouel commented on pkg/adapter/incoming.go:
```diff
@@ -86,6 +121,97 @@
+	ref: https://docs.github.com/en/developers/apps/building-github-apps/authenticating-with-github-apps#authenticating-as-an-installation ,
+	     https://docs.github.com/en/rest/apps/installations?apiVersion=2022-11-28#list-repositories-accessible-to-the-app-installation */
+	for i := range installationData {
+		res, err := getResponse(ctx, http.MethodPost, fmt.Sprintf(installationTokenURL, *installationData[i].ID), JWTtoken, run)
```
the installation token is going to bound to every repository which could be a security issue since this would include private repos in a org. look at how it's done in parse_payload to make sure it's bounded to the repo where it's coming from unless the configuration setting is set (and please use go-github library for that)

savitaashture replied:
To get repository list used go-github library instead of API call

chmouel commented on pkg/adapter/incoming_test.go:
```diff
@@ -21,10 +23,27 @@
 	rtesting "knative.dev/pkg/reconciler/testing"
 )
 
+const fakePrivateKey = `-----BEGIN RSA PRIVATE KEY-----
```
not our security team robot is going to bug us when they sees this 🙄

savitaashture replied:
its just fake data 
does it create issue :thinking:

jkandasa replied:
It is not a good advice to keep sensitive data in source, even if it is a fake data.
We have to give explanation each time when it hits security audit.
better we can generate it dynamically

chmouel commented on docs/content/docs/guide/incoming_webhook.md:
```diff
@@ -14,10 +14,9 @@ specify a reference Secret which will be used as a shared secret and the
-the [github webhook documentation](/docs/install/github_webhook) for how to set
-this up.
+you will need to have a `git_provider` spec to specify a token when using the
+github-webhook method. Refer to the [github webhook documentation](/docs/install/github_webhook)
```
a non `github app` ? because it's not only github-webhook

savitaashture replied:
Updated doc

chmouel reviewed (commented):
Good start, a few comments mainly about code duplication and scopped token

savitaashture commented:
/retest

sm43 commented:
/retest

chetan-rns commented:
/retest

chmouel commented on docs/content/docs/guide/incoming_webhook.md:
```diff
@@ -14,10 +14,10 @@ specify a reference Secret which will be used as a shared secret and the
-github-apps method the same way we are doing for github-webhook method. Refer to
-the [github webhook documentation](/docs/install/github_webhook) for how to set
-this up.
+you will need to have a `git_provider` spec to specify a token when using a non github app.
```
if you are not using the github app provider (ie: webhook based provider) you will need to have to specify a token.

chmouel commented on docs/content/docs/guide/incoming_webhook.md:
```diff
@@ -27,8 +27,12 @@ values are:
 - gitlab
 - bitbucket-cloud
 
+Whereas for `github-apps` we detect secret if GithubApp is installed on the cluster so no need to specify a `git_provider` spec.
```
Whereas for `github-apps` this doesn't need to be added

chmouel commented on pkg/adapter/incoming_test.go:
```diff
@@ -63,6 +82,13 @@ func Test_compareSecret(t *testing.T) {
+		},
+	}
+
 	type args struct {
```
use gotest env to recover the env after

chmouel commented on pkg/adapter/incoming_test.go:
```diff
@@ -536,3 +609,173 @@ func Test_listener_processIncoming(t *testing.T) {
+}
+
+func Test_GetAndUpdateInstallationID(t *testing.T) {
+	os.Setenv("SYSTEM_NAMESPACE", "pipelinesascode")
```
same comment as before,

chmouel commented on pkg/adapter/incoming_test.go:
```diff
@@ -536,3 +609,173 @@ func Test_listener_processIncoming(t *testing.T) {
+		},
+	}
+
+	os.Setenv("SYSTEM_NAMESPACE", "pipelinesascode")
```
gotest.env

chmouel commented on pkg/adapter/incoming.go:
```diff
@@ -19,6 +26,13 @@ import (
 
+var (
+	// default is "https://api.github.com" but it can be overridden by X-GitHub-Enterprise-Host header
+	apiURL = "https://api.github.com"
```
we should move that to the keys package

savitaashture replied:
Moved

chmouel commented on pkg/adapter/incoming.go:
```diff
@@ -68,8 +82,20 @@ func (l *listener) detectIncoming(ctx context.Context, req *http.Request, payloa
+		}
+		// Github app is not installed for provided repository url
+		if l.event.InstallationID == 0 {
+			return false, nil, fmt.Errorf("GithubApp is not installed for provided repository url %s ", repo.Spec.URL)
```
for the provided

chmouel commented on docs/content/docs/guide/incoming_webhook.md:
````diff
@@ -52,7 +56,28 @@ spec:
+      type: webhook-url
+```
+
+a secret named `repo-incoming-secret` for both webhook and github-apps will have this value:
````
maybe fix the typo on line 94 for piplinerun1

chmouel commented on pkg/adapter/incoming.go:
```diff
@@ -86,6 +112,78 @@ func (l *listener) detectIncoming(ctx context.Context, req *http.Request, payloa
+	ref: https://docs.github.com/en/developers/apps/building-github-apps/authenticating-with-github-apps#authenticating-as-an-installation ,
+	     https://docs.github.com/en/rest/apps/installations?apiVersion=2022-11-28#list-repositories-accessible-to-the-app-installation */
+	for i := range installationData {
+		l.event.Provider.URL = enterpriseURL
```
isn't that a bug ? and you don't have fallback in the previous logic to the public url

savitaashture replied:
fixed that 

```
l.event.Provider.URL = enterpriseURL
```

if `l.event.Provider.URL` is empty then it will use https://github.com/openshift-pipelines/pipelines-as-code/blob/main/pkg/provider/github/parse_payload.go#L61-L63 public one

chmouel reviewed (commented):
some reviews, and we really need e2e test to cover this, a gitlab and a github  app one

savitaashture committed:
Added e2e test to verify Incoming for GithubApp

savitaashture commented:
> some reviews, and we really need e2e test to cover this, a gitlab and a github app one

@chmouel we have e2e test for Gitlab here https://github.com/openshift-pipelines/pipelines-as-code/blob/main/test/gitlab_incoming_webhook_test.go

So added e2e test for GithubApp now for Incoming

chmouel commented:
lmk when you finish added the e2e and i'll do a new review

savitaashture commented:
> lmk when you finish added the e2e and i'll do a new review

https://github.com/openshift-pipelines/pipelines-as-code/pull/1090#issuecomment-1377404192 I have added e2e already

savitaashture commented:
The e2e test not failing locally :thinking:

savitaashture committed:
Enhance incoming webhook for github apps
If GithubApp is installed on the cluster then no need to provide git_provider spec because for GithubApps we detect the secret and use it

Signed-off-by: Savita Ashture <sashture@redhat.com>

savitaashture commented:
/retest

savitaashture commented:
/test

chmouel commented:
/retest

chmouel commented:
glad we have pac running :-p  unitlint is broken...

chmouel commented on pkg/adapter/incoming.go:
```diff
@@ -86,6 +106,79 @@ func (l *listener) detectIncoming(ctx context.Context, req *http.Request, payloa
+				return err
+			}
+		}
+		repoList, err := gh.ListRepos(ctx)
```
let's move this to its own repo, add some paginations in a further iteration and add some comment at top to specify it list all repos belonging to the tokens

chmouel commented on pkg/adapter/incoming.go:
```diff
@@ -95,20 +188,65 @@ func (l *listener) processIncoming(targetRepo *v1alpha1.Repository) (provider.In
+	jwt.RegisteredClaims
+}
+
+func GenerateJWT(ctx context.Context, run *params.Run) (string, error) {
```
let's see if we can move this functions that are specific to github inside the github package,

chmouel commented on pkg/adapter/incoming_test.go:
```diff
@@ -63,6 +83,13 @@ func Test_compareSecret(t *testing.T) {
 
 func Test_listener_detectIncoming(t *testing.T) {
 	const goodURL = "https://matched/by/incoming"
+	t.Setenv("SYSTEM_NAMESPACE", "pipelinesascode")
```
defer env.Patch|("SYSTEM_NAMESPACE", "pipelineascode")()
https://github.com/openshift-pipelines/pipelines-as-code/blob/update-to-v1/test/gitea_test.go#L542

jkandasa replied:
@chmouel I see `env.Patch` is deprecated, and asked to use `t.Setenv`
https://pkg.go.dev/gotest.tools/v3@v3.4.0/env#Patch

savitaashture replied:
Yes right

We have another one which is `env.PatchAll` we can use that

savitaashture replied:
Used t.Setenv

chmouel commented on pkg/adapter/incoming_test.go:
```diff
@@ -536,3 +610,173 @@ func Test_listener_processIncoming(t *testing.T) {
+		},
+	}
+
+	t.Setenv("SYSTEM_NAMESPACE", "pipelinesascode")
```
do the same as before

jkandasa commented on pkg/adapter/incoming.go:
```diff
@@ -68,8 +76,20 @@ func (l *listener) detectIncoming(ctx context.Context, req *http.Request, payloa
+		if l.event.InstallationID == 0 {
+			return false, nil, fmt.Errorf("GithubApp is not installed for the provided repository url %s ", repo.Spec.URL)
+		}
 	}
```
make this block simple as,
```
	if repo.Spec.GitProvider == nil ||  repo.Spec.GitProvider.Type == "" {
		if err = l.getAndUpdateInstallationID(ctx, req, l.run, repo); err != nil {
			return false, nil, err
		}
		// Github app is not installed for provided repository url
		if l.event.InstallationID == 0 {
			return false, nil, fmt.Errorf("GithubApp is not installed for the provided repository url %s ", repo.Spec.URL)
		}
	}
```

savitaashture replied:
Updated

jkandasa commented on pkg/adapter/incoming.go:
```diff
@@ -68,8 +76,20 @@ func (l *listener) detectIncoming(ctx context.Context, req *http.Request, payloa
+		isProviderTypeNotSet = true
+	}
+	if isProviderTypeNotSet {
+		if err = l.getAndUpdateInstallationID(ctx, req, l.run, repo); err != nil {
```
receiver function `l.getAndUpdateInstallationID` can access `l.run` itself, is there a reason to pass it from here?

savitaashture replied:
Refactored code and now it won't have receiver type

jkandasa commented on pkg/adapter/incoming.go:
```diff
@@ -86,6 +106,79 @@ func (l *listener) detectIncoming(ctx context.Context, req *http.Request, payloa
+	for i := range installationData {
+		l.event.Provider.URL = enterpriseURL
+		gh := github.New()
+		if *installationData[i].ID != 0 {
```
if the `ID` is `nil`, leads panic

savitaashture replied:
:thinking: 
I am accessing *installationData[i].ID when there is installationData 
So i did not face this panic issue 
will it lead panic ?

jkandasa replied:
`ID` is a pointer, can be `nil`
https://go.dev/play/p/Rirp8FkYtNh

savitaashture replied:
Oh ya there can be panic issue
As installation ID was never nil so did not thought about that

But added condition to make sure that scenario

jkandasa commented on pkg/adapter/incoming.go:
```diff
@@ -95,20 +188,65 @@ func (l *listener) processIncoming(targetRepo *v1alpha1.Repository) (provider.In
-	switch targetRepo.Spec.GitProvider.Type {
-	case "github":
+	if isProviderTypeNotSet {
 		provider = github.New()
```
make this block simple as,
```
var provider provider.Interface
if targetRepo.Spec.GitProvider == nil || targetRepo.Spec.GitProvider.Type == "" {
    provider = github.New()
} else {
```

savitaashture replied:
Updated

jkandasa commented on pkg/adapter/incoming.go:
```diff
@@ -95,20 +188,65 @@ func (l *listener) processIncoming(targetRepo *v1alpha1.Repository) (provider.In
+		case "bitbucket-cloud":
+			provider = &bitbucketcloud.Provider{}
+		case "bitbucket-server":
+			provider = &bitbucketserver.Provider{}
```
I think it is better to have a custom GitProvider type from string and consuming it across.
keeping raw string may lead maintainability issue in long term

savitaashture replied:
Yes agree

With PAC we are supporting only those VCS thats why hard coded as of now
But will consider this point if there is anything newly added

savitaashture committed:
Updated incoming webhook implementation as per the review comments

jkandasa commented on pkg/provider/github/app/token.go:
```diff
@@ -0,0 +1,131 @@
+		return "", err
+	}
+
+	expirationTime := time.Now().Add(5 * time.Minute)
```
do we have any note, why we use 5 minutes as `expirationTime`. If non at least add a note here.

savitaashture replied:
Added note :+1

chmouel commented:
you are right @jkandasa about t.Setenv, i think we will then need a refactoring of all our test using gotest.tools env to move to t.Setenv

savitaashture committed:
Add test to cover ListRepos and fixed review comments

savitaashture commented:
/retest

chmouel committed:
Merge branch 'main' into enhance_incoming_webhook
